README.md
# richterm

![](https://github.com/Abdur-rahmaanJ/richterm/blob/stable/richterm.gif)


```rust
use richterm::print;
use richterm::text;
use richterm::track;
use richterm::progress;

use std::{thread, time};

fn main() {
    let vec = vec![text(
        "richterm",
        "fg:dark_blue bg:indian_red1 eff:b,i,u,blink,s",
    )];
    print(vec);

    let arr = [text(" supports ", "")];
    print(arr);

    print([
        text(" many ", "bg:sea_green2 fg:black eff:i"),
        text(" features ", ""),
        text(" out of the box", "bg:yellow fg:black"),
        text("\n", ""),
    ]);

    print([
        text("Even", ""),
        text(" some ", "eff:b"),
        text(
            "emojis like :+1: :apple: :bar_chart: :airplane_departure: :baguette_bread: :minibus:",
            "",
        ),
        text("\n", ""),
    ]);

    for _item in track(5, "Downloading ...") {
        let ten_millis = time::Duration::from_millis(5);
        thread::sleep(ten_millis);
    }
    
    let mut tasks = progress();
    let task1 = tasks.add_task("Descr1", 50);
    let task2 = tasks.add_task("Descr2", 50); 
    let task3 = tasks.add_task("Descr3", 50);
    
    while !tasks.finished{
        tasks.update(task1.clone(), 10.0);
        tasks.update(task2.clone(), 5.0);
        tasks.update(task3.clone(), 6.0);
    } 
}

```


# Docs


### fg and bg colors

- rgb(111,111,111)
- black, red, green, yellow, blue, magenta, cyan, white, bright_black, bright_red, bright_green, bright_yellow, bright_blue, bright_magenta, bright_cyan, bright_white, grey0, gray0, navy_blue, dark_blue, blue3, blue1, dark_green, deep_sky_blue4, dodger_blue3, dodger_blue2, green4, spring_green4, turquoise4, deep_sky_blue3, dodger_blue1, green3, spring_green3, dark_cyan, light_sea_green, deep_sky_blue2, deep_sky_blue1, spring_green2, cyan3, dark_turquoise, turquoise2, green1, spring_green1, medium_spring_green, cyan2, cyan1, dark_red, deep_pink4, purple4, purple3, blue_violet, orange4, grey37, gray37, medium_purple4, slate_blue3, royal_blue1, chartreuse4, dark_sea_green4, pale_turquoise4, steel_blue, steel_blue3, cornflower_blue, chartreuse3, cadet_blue, sky_blue3, steel_blue1, pale_green3, sea_green3, aquamarine3, medium_turquoise, chartreuse2, sea_green2, sea_green1, aquamarine1, dark_slate_gray2, dark_magenta, dark_violet, purple, light_pink4, plum4, medium_purple3, slate_blue1, yellow4, wheat4, grey53, gray53, light_slate_grey, light_slate_gray, medium_purple, light_slate_blue, dark_olive_green3, dark_sea_green, light_sky_blue3, sky_blue2, dark_sea_green3, dark_slate_gray3, sky_blue1, chartreuse1, light_green, pale_green1, dark_slate_gray1, red3, medium_violet_red, magenta3, dark_orange3, indian_red, hot_pink3, medium_orchid3, medium_orchid, medium_purple2, dark_goldenrod, light_salmon3, rosy_brown, grey63, gray63, medium_purple1, gold3, dark_khaki, navajo_white3, grey69, gray69, light_steel_blue3, light_steel_blue, yellow3, dark_sea_green2, light_cyan3, light_sky_blue1, green_yellow, dark_olive_green2, dark_sea_green1, pale_turquoise1, deep_pink3, magenta2, hot_pink2, orchid, medium_orchid1, orange3, light_pink3, pink3, plum3, violet, light_goldenrod3, tan, misty_rose3, thistle3, plum2, khaki3, light_goldenrod2, light_yellow3, grey84, gray84, light_steel_blue1, yellow2, dark_olive_green1, honeydew2, light_cyan1, red1, deep_pink2, deep_pink1, magenta1, orange_red1, indian_red1, hot_pink, dark_orange, salmon1, light_coral, pale_violet_red1, orchid2, orchid1, orange1, sandy_brown, light_salmon1, light_pink1, pink1, plum1, gold1, navajo_white1, misty_rose1, thistle1, yellow1, light_goldenrod1, khaki1, wheat1, cornsilk1, grey100, gray100, grey3, gray3, grey7, gray7, grey11, gray11, grey15, gray15, grey19, gray19, grey23, gray23, grey27, gray27, grey30, gray30, grey35, gray35, grey39, gray39, grey42, gray42, grey46, gray46, grey50, gray50, grey54, gray54, grey58, gray58, grey62, gray62, grey66, gray66, grey70, gray70, grey74, gray74, grey78, gray78, grey82, gray82, grey85, gray85, grey89, gray89, grey93, gray93, 


### effects

- b - bold
- i - italic
- u - underline
- blink - blink
- d - dim
- r - reverse

### emojis

Try github-flavoured shortcodes

### track(5, "description")

Progressbar

### progress

```rust
    let mut tasks = progress();
    let task1 = tasks.add_task("Descr1", 50); // 50 is an arbitrary amount
    let task2 = tasks.add_task("Descr2", 50); 
    let task3 = tasks.add_task("Descr3", 50);
    
    while !tasks.finished{
        tasks.update(task1.clone(), 10.0);  // 10.0 is an arbitrary amount to move the progress bar
        tasks.update(task2.clone(), 5.0);   // 10.0 means 10.0/50.0
        tasks.update(task3.clone(), 6.0);
    } 
```


CHANGELOG.md
0.5.0

- feat: progress
- progress and progressbar use structs

0.4.0

- track/Progressbar

0.3.0

- Print accepts arrays and vectors
- Emoji shortcodes

0.2.0

- Optional positioning of args 
- effects can be combined


src/lib.rs
mod colors;
mod emojis;
mod uuid;
mod spinners;
use regex::escape;
use regex::Regex;
use std::collections::HashMap;
use std::io::{self, Write};

pub fn print<T: AsRef<[U]>, U: std::fmt::Debug + std::fmt::Display>(data: T) {
    let slice: &[U] = data.as_ref();

    for item in slice {
        // Process the elements here
        print!("{}", item);
        io::stdout().flush().unwrap();
    }
}

// Function to replace emoji shortcodes in a given text
fn replace_emoji_shortcodes(text: &str) -> String {
    let emoji_map = emojis::emoji_shortcodes();
    let mut replaced_text = String::from(text);

    for (shortcode, emoji) in emoji_map.iter() {
        let regex_str = format!(":{}:", escape(shortcode));
        let regex = Regex::new(&regex_str).unwrap();

        replaced_text = regex.replace_all(&replaced_text, *emoji).to_string();
    }

    replaced_text
}

// fg:100 bg:200 b:0 eff:i
pub fn text(usertext: &str, format: &str) -> String {
    // basic checks
    if format.is_empty() {
        let retstr = replace_emoji_shortcodes(usertext);
        return retstr;
    }

    let escape = "\x1b";
    let end_seq = "\x1b[0m";
    let mut all_effects = String::from("");
    let ansi_codes = colors::ansi_color_codes();

    let parts: Vec<&str> = format.split_whitespace().collect();
    for part in parts {
        let part_all: Vec<&str> = part.split(':').collect(); // ex fg:red
        let command = part_all[0]; // ex fg
        let val = part_all[1]; // ex red

        if command == "fg" {
            if val.starts_with("rgb(") {
                let trimmed_start = val.trim_start_matches("rgb(");
                let trimmed_end = trimmed_start.trim_end_matches(")");
                let rgb: Vec<&str> = trimmed_end.split(',').collect();
                let rgb_formatted = format!("{}[38;2;{};{};{}m", escape, rgb[0], rgb[1], rgb[2]);
                all_effects.push_str(&rgb_formatted);
            } else {
                let fg_code = ansi_codes
                    .get(&val)
                    .expect("Cannot find color name")
                    .to_string();
                let fg_formatted = format!("{}[38;5;{}m", escape, fg_code);
                all_effects.push_str(&fg_formatted);
            }
        } else if command == "bg" {
            if val.starts_with("rgb(") {
                let trimmed_start = val.trim_start_matches("rgb(");
                let trimmed_end = trimmed_start.trim_end_matches(')');
                let rgb: Vec<&str> = trimmed_end.split(',').collect();
                let rgb_formatted = format!("{}[48;2;{};{};{}m", escape, rgb[0], rgb[1], rgb[2]);
                all_effects.push_str(&rgb_formatted);
            } else {
                let bg_code = ansi_codes
                    .get(&val)
                    .expect("Cannot find color name")
                    .to_string();
                let bg_formatted = format!("{}[48;5;{}m", escape, bg_code);
                all_effects.push_str(&bg_formatted);
            }
        } else if command == "eff" {
            let user_effects: Vec<&str> = val.split(',').collect();
            for effect in user_effects {
                match effect {
                    "i" => {
                        let e = format!("{}[3m", escape);
                        all_effects.push_str(&e);
                    }
                    "blink" => {
                        let e = format!("{}[5m", escape);
                        all_effects.push_str(&e);
                    }
                    "u" => {
                        let e = format!("{}[4m", escape);
                        all_effects.push_str(&e);
                    }
                    "b" => {
                        let e = format!("{}[1m", escape);
                        all_effects.push_str(&e);
                    }
                    "s" => {
                        let e = format!("{}[9m", escape);
                        all_effects.push_str(&e);
                    }
                    "d" => {
                        let e = format!("{}[2m", escape);
                        all_effects.push_str(&e);
                    }
                    "r" => {
                        let e = format!("{}[7m", escape);
                        all_effects.push_str(&e);
                    }
                    "hc" => {
                        let e = format!("{}[?25l", escape);
                        all_effects.push_str(&e);
                    }
                    "sc" => {
                        let e = format!("{}[?25h", escape);
                        all_effects.push_str(&e);
                    }
                    &_ => {
                        all_effects.push_str("");
                    }
                }
            }
        }
    }

    let usert = replace_emoji_shortcodes(usertext);
    // TODO: emoji
    let to_ret = format!("{}{}{}", all_effects, usert, end_seq);

    to_ret
}

// === BrogressBar ===
trait Show {
    fn show(&self);
}

trait ShowCompleted {
    fn show_completed(&self);
}

// We at richterm promote the community.
// Pro-gress implies individualistic excellence,
// which denotes a group of elites in their craft and
// average people.
// This is why we at richterm promote bro-gress.
// We operate, despite the  differences in levels,
// as siblings. This increases the rate of achieving the
// targetted objective drastically.
// Note: Summary for the lazy:
//   We simply operate as bros, rather than pros.
#[derive(Debug, Clone)]
pub struct BrogressBar {
    description: String,
    steps: i32,
    steps_done: f32,
    start_time: Instant,
    display: String,
    bar_width: i32,
    time_taken_f: String,
    is_finished: bool,
}

impl BrogressBar {
    pub fn new(steps: i32, description: &str) -> BrogressBar {
        BrogressBar {
            description: description.to_string(),
            steps: steps,
            steps_done: 0.0,
            start_time: Instant::now(),
            display: "━".to_string(),
            bar_width: 40,
            time_taken_f: "".to_string(),
            is_finished: false,
        }
    }

    pub fn update(&mut self, step: f32) {
        let elapsed_time = self.start_time.elapsed();
        let elapsed_time_f = format!("{} ", format_duration(elapsed_time));

        if (self.steps_done + step) >= self.steps as f32 {
            self.steps_done = self.steps as f32;
            self.is_finished = true;
            if self.time_taken_f == "" {
                self.time_taken_f = elapsed_time_f;
            }
        } else {
            self.steps_done = self.steps_done + step;
        }
    }

    pub fn finished(&self) -> bool {
        if (self.steps_done / self.steps as f32) == 1.0 {
            return true;
        }
        return false;
    }

    pub fn get_show_texts(&self) -> Vec<String> {
        let perc = (self.steps_done as f64 / self.steps as f64) * 100.0;
        let done = ((self.steps_done as f64 / self.steps as f64) * self.bar_width as f64) as i32;
        let undone = self.bar_width - done;
        let elapsed_time = self.start_time.elapsed();
        let elapsed_time_f = format!("{} ", format_duration(elapsed_time));

        let mut v = vec![];
        if self.is_finished {
            v.extend(vec![
                text("\r ", ""),
                text(&self.description, ""),
                text(" ", ""),
                text(
                    &self.display.repeat(self.bar_width.try_into().unwrap()),
                    "fg:green",
                ),
                text(" 100% ", "fg:magenta"),
                text(&self.time_taken_f, "fg:orange1"),
            ]);
        } else {
            v.extend(vec![
                text("\r ", "eff:hc"),
                text(&self.description, ""),
                text(" ", ""),
                text(
                    &self.display.repeat(done.try_into().unwrap()),
                    "fg:deep_pink3",
                ),
                text(&self.display.repeat(undone.try_into().unwrap()), "fg:black"),
                text(&format!(" {}% ", perc as i32), "fg:magenta"),
                text(&elapsed_time_f, "fg:steel_blue eff:sc"),
            ]);
        }

        v
    }
}

impl Show for BrogressBar {
    fn show(&self) {
        print([
            text("\r ", ""),
            text(&self.description, ""),
            text(" ", ""),
            text(
                &self.display.repeat(self.bar_width.try_into().unwrap()),
                "fg:green",
            ),
            text(" 100% ", "fg:magenta"),
            text(&self.time_taken_f, "fg:orange1"),
        ]);
    }
}

impl ShowCompleted for BrogressBar {
    fn show_completed(&self) {
        print([
            text("\r ", ""),
            text(&self.description, ""),
            text(" ", ""),
            text(
                &self.display.repeat(self.bar_width.try_into().unwrap()),
                "fg:green",
            ),
            text(" 100% ", "fg:magenta"),
            text(&self.time_taken_f, "fg:orange1"),
            text("\n", ""),
        ]);
    }
}

use std::iter::Iterator;

// Custom Iterator struct
pub struct MyRange {
    start: i32,
    end: i32,
    brogress_bar: BrogressBar,
}

// Implement the Iterator trait for MyRange
impl Iterator for MyRange {
    type Item = i32;

    fn next(&mut self) -> Option<Self::Item> {
        if self.start <= self.end {
            let current = self.start;
            self.start += 1;
            Some(current)
        } else {
            None
        }
    }
}

impl Drop for MyRange {
    fn drop(&mut self) {
        // Code to be executed after the last iteration.
        self.brogress_bar.show_completed();
    }
}

// Idk what i wrote here
fn create_range<'a>(start: i32, end: i32, brogress_bar: BrogressBar) -> MyRange {
    MyRange {
        start,
        end,
        brogress_bar,
    }
}

use std::time::{Duration, Instant};

fn format_duration(duration: Duration) -> String {
    let hours = duration.as_secs() / 3600;
    let minutes = (duration.as_secs() % 3600) / 60;
    let seconds = duration.as_secs() % 60;

    format!("{:02}:{:02}:{:02}", hours, minutes, seconds)
}

#[derive(Clone)]
pub struct Brogress {
    pub finished: bool,
    tasks: HashMap<String, BrogressBar>,
    inserts_order: Vec<String>,
}

impl Brogress {
    fn new() -> Self {
        Brogress {
            finished: false,
            tasks: HashMap::new(),
            inserts_order: Vec::new(),
        }
    }

    pub fn add_task(&mut self, description: &str, steps: i32) -> String {
        let x = uuid::generate_random_uuid();
        let brogress_bar = BrogressBar::new(steps, description);
        self.tasks.insert(x.clone(), brogress_bar);
        self.inserts_order.push(x.clone());
        x.to_string()
    }

    pub fn update(&mut self, uuid: String, step: f32) {
        if self.finished {
            print([
                text("", "eff:sc"),
                text(&format!("\x1B[{}B", self.tasks.len()), ""),
            ]);
            return;
        }
        if !self.tasks.contains_key(&uuid) {
            println!("Cannot find tasks");
        }

        // let brogress_bar = self.tasks.get_mut(uuid).as_mut();
        let Some(brogress_bar) = self.tasks.get_mut(&uuid) else { todo!() };
        brogress_bar.update(step);
        // brogress_bar.show();
        // let moveup = "\x1B[1F";
        // let movedown = "\x1B[2B";
        // let repeated_chars: String = std::iter::repeat(moveuperase).take(self.tasks.len()).collect();
        let mut sum_finished = 0;
        // let mut index = 0;

        let mut to_print = vec![text("", "eff:hc")];

        for uuid in &self.inserts_order {
            let brogress_bar = self.tasks.get(uuid);
            if brogress_bar.expect("REASON").finished() {
                sum_finished += 1;
            }
            // brogress_bar.expect("REASON").show();
            to_print.extend(brogress_bar.expect("REASON").get_show_texts());
            to_print.push("\n".to_string());
        }

        print(to_print);

        print!("\x1B[{}F", self.tasks.len());

        if sum_finished == self.tasks.len() {
            self.finished = true;
        }
    }
}

// impl Drop for Brogress {
//     fn drop(&mut self) {
//         // This block will be executed when the object goes out of scope
//         println!("Cleaning up resources for data: {}", self.data);
//         // Perform any necessary cleanup or resource release here
//     }
// }

pub fn track(steps: i32, description: &str) -> MyRange {
    // Downloading ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:21

    let mut brogress_bar = BrogressBar::new(steps, description);
    for _item in 1..steps {
        brogress_bar.update(1.0);
        brogress_bar.show();
    }
    return create_range(1, steps, brogress_bar);
}

pub fn progress() -> Brogress {
    let p = Brogress::new();

    p
}

pub fn spinner(type_: &str, message: &str) -> spinners::Spinner{
    let s = spinners::Spinner::new(type_, message);

    s
}

src/colors.rs
use convert_case::{Case, Casing};
use std::str::FromStr;
use strum::EnumString;

#[derive(Debug, PartialEq, EnumString)]
enum Colors {
    #[strum(ascii_case_insensitive)]
    Black = 0,
    #[strum(ascii_case_insensitive)]
    Red = 1,
    #[strum(ascii_case_insensitive)]
    Green = 2,
    #[strum(ascii_case_insensitive)]
    Yellow = 3,
    #[strum(ascii_case_insensitive)]
    Blue = 4,
    #[strum(ascii_case_insensitive)]
    Magenta = 5,
    #[strum(ascii_case_insensitive)]
    Cyan = 6,
    #[strum(ascii_case_insensitive)]
    White = 7,
    #[strum(ascii_case_insensitive)]
    BrightBlack = 8,
    #[strum(ascii_case_insensitive)]
    BrightRed = 9,
    #[strum(ascii_case_insensitive)]
    BrightGreen = 10,
    #[strum(ascii_case_insensitive)]
    BrightYellow = 11,
    #[strum(ascii_case_insensitive)]
    BrightBlue = 12,
    #[strum(ascii_case_insensitive)]
    BrightMagenta = 13,
    #[strum(ascii_case_insensitive)]
    BrightCyan = 14,
    #[strum(ascii_case_insensitive)]
    BrightWhite = 15,
    #[strum(ascii_case_insensitive)]
    Grey0 = 16,
    #[strum(ascii_case_insensitive)]
    NavyBlue = 17,
    #[strum(ascii_case_insensitive)]
    DarkBlue = 18,
    #[strum(ascii_case_insensitive)]
    Blue3 = 20,
    #[strum(ascii_case_insensitive)]
    Blue1 = 21,
    #[strum(ascii_case_insensitive)]
    DarkGreen = 22,
    #[strum(ascii_case_insensitive)]
    DeepSkyBlue4 = 25,
    #[strum(ascii_case_insensitive)]
    DodgerBlue3 = 26,
    #[strum(ascii_case_insensitive)]
    DodgerBlue2 = 27,
    #[strum(ascii_case_insensitive)]
    Green4 = 28,
    #[strum(ascii_case_insensitive)]
    SpringGreen4 = 29,
    #[strum(ascii_case_insensitive)]
    Turquoise4 = 30,
    #[strum(ascii_case_insensitive)]
    DeepSkyBlue3 = 32,
    #[strum(ascii_case_insensitive)]
    DodgerBlue1 = 33,
    #[strum(ascii_case_insensitive)]
    Green3 = 40,
    #[strum(ascii_case_insensitive)]
    SpringGreen3 = 41,
    #[strum(ascii_case_insensitive)]
    DarkCyan = 36,
    #[strum(ascii_case_insensitive)]
    LightSeaGreen = 37,
    #[strum(ascii_case_insensitive)]
    DeepSkyBlue2 = 38,
    #[strum(ascii_case_insensitive)]
    DeepSkyBlue1 = 39,
    #[strum(ascii_case_insensitive)]
    SpringGreen2 = 47,
    #[strum(ascii_case_insensitive)]
    Cyan3 = 43,
    #[strum(ascii_case_insensitive)]
    DarkTurquoise = 44,
    #[strum(ascii_case_insensitive)]
    Turquoise2 = 45,
    #[strum(ascii_case_insensitive)]
    Green1 = 46,
    #[strum(ascii_case_insensitive)]
    SpringGreen1 = 48,
    #[strum(ascii_case_insensitive)]
    MediumSpringGreen = 49,
    #[strum(ascii_case_insensitive)]
    Cyan2 = 50,
    #[strum(ascii_case_insensitive)]
    Cyan1 = 51,
    #[strum(ascii_case_insensitive)]
    DarkRed = 88,
    #[strum(ascii_case_insensitive)]
    DeepPink4 = 125,
    #[strum(ascii_case_insensitive)]
    Purple4 = 55,
    #[strum(ascii_case_insensitive)]
    Purple3 = 56,
    #[strum(ascii_case_insensitive)]
    BlueViolet = 57,
    #[strum(ascii_case_insensitive)]
    Orange4 = 94,
    #[strum(ascii_case_insensitive)]
    Grey37 = 59,
    #[strum(ascii_case_insensitive)]
    MediumPurple4 = 60,
    #[strum(ascii_case_insensitive)]
    SlateBlue3 = 62,
    #[strum(ascii_case_insensitive)]
    RoyalBlue1 = 63,
    #[strum(ascii_case_insensitive)]
    Chartreuse4 = 64,
    #[strum(ascii_case_insensitive)]
    DarkSeaGreen4 = 71,
    #[strum(ascii_case_insensitive)]
    PaleTurquoise4 = 66,
    #[strum(ascii_case_insensitive)]
    SteelBlue = 67,
    #[strum(ascii_case_insensitive)]
    SteelBlue3 = 68,
    #[strum(ascii_case_insensitive)]
    CornflowerBlue = 69,
    #[strum(ascii_case_insensitive)]
    Chartreuse3 = 76,
    #[strum(ascii_case_insensitive)]
    CadetBlue = 73,
    #[strum(ascii_case_insensitive)]
    SkyBlue3 = 74,
    #[strum(ascii_case_insensitive)]
    SteelBlue1 = 81,
    #[strum(ascii_case_insensitive)]
    PaleGreen3 = 114,
    #[strum(ascii_case_insensitive)]
    SeaGreen3 = 78,
    #[strum(ascii_case_insensitive)]
    Aquamarine3 = 79,
    #[strum(ascii_case_insensitive)]
    MediumTurquoise = 80,
    #[strum(ascii_case_insensitive)]
    Chartreuse2 = 112,
    #[strum(ascii_case_insensitive)]
    SeaGreen2 = 83,
    #[strum(ascii_case_insensitive)]
    SeaGreen1 = 85,
    #[strum(ascii_case_insensitive)]
    Aquamarine1 = 122,
    #[strum(ascii_case_insensitive)]
    DarkSlateGray2 = 87,
    #[strum(ascii_case_insensitive)]
    DarkMagenta = 91,
    #[strum(ascii_case_insensitive)]
    DarkViolet = 128,
    #[strum(ascii_case_insensitive)]
    Purple = 129,
    #[strum(ascii_case_insensitive)]
    LightPink4 = 95,
    #[strum(ascii_case_insensitive)]
    Plum4 = 96,
    #[strum(ascii_case_insensitive)]
    MediumPurple3 = 98,
    #[strum(ascii_case_insensitive)]
    SlateBlue1 = 99,
    #[strum(ascii_case_insensitive)]
    Yellow4 = 106,
    #[strum(ascii_case_insensitive)]
    Wheat4 = 101,
    #[strum(ascii_case_insensitive)]
    Grey53 = 102,
    #[strum(ascii_case_insensitive)]
    LightSlateGrey = 103,
    #[strum(ascii_case_insensitive)]
    MediumPurple = 104,
    #[strum(ascii_case_insensitive)]
    LightSlateBlue = 105,
    #[strum(ascii_case_insensitive)]
    DarkOliveGreen3 = 149,
    #[strum(ascii_case_insensitive)]
    DarkSeaGreen = 108,
    #[strum(ascii_case_insensitive)]
    LightSkyBlue3 = 110,
    #[strum(ascii_case_insensitive)]
    SkyBlue2 = 111,
    #[strum(ascii_case_insensitive)]
    DarkSeaGreen3 = 150,
    #[strum(ascii_case_insensitive)]
    DarkSlateGray3 = 116,
    #[strum(ascii_case_insensitive)]
    SkyBlue1 = 117,
    #[strum(ascii_case_insensitive)]
    Chartreuse1 = 118,
    #[strum(ascii_case_insensitive)]
    LightGreen = 120,
    #[strum(ascii_case_insensitive)]
    PaleGreen1 = 156,
    #[strum(ascii_case_insensitive)]
    DarkSlateGray1 = 123,
    #[strum(ascii_case_insensitive)]
    Red3 = 160,
    #[strum(ascii_case_insensitive)]
    MediumVioletRed = 126,
    #[strum(ascii_case_insensitive)]
    Magenta3 = 164,
    #[strum(ascii_case_insensitive)]
    DarkOrange3 = 166,
    #[strum(ascii_case_insensitive)]
    IndianRed = 167,
    #[strum(ascii_case_insensitive)]
    HotPink3 = 168,
    #[strum(ascii_case_insensitive)]
    MediumOrchid3 = 133,
    #[strum(ascii_case_insensitive)]
    MediumOrchid = 134,
    #[strum(ascii_case_insensitive)]
    MediumPurple2 = 140,
    #[strum(ascii_case_insensitive)]
    DarkGoldenrod = 136,
    #[strum(ascii_case_insensitive)]
    LightSalmon3 = 173,
    #[strum(ascii_case_insensitive)]
    RosyBrown = 138,
    #[strum(ascii_case_insensitive)]
    Grey63 = 139,
    #[strum(ascii_case_insensitive)]
    MediumPurple1 = 141,
    #[strum(ascii_case_insensitive)]
    Gold3 = 178,
    #[strum(ascii_case_insensitive)]
    DarkKhaki = 143,
    #[strum(ascii_case_insensitive)]
    NavajoWhite3 = 144,
    #[strum(ascii_case_insensitive)]
    Grey69 = 145,
    #[strum(ascii_case_insensitive)]
    LightSteelBlue3 = 146,
    #[strum(ascii_case_insensitive)]
    LightSteelBlue = 147,
    #[strum(ascii_case_insensitive)]
    Yellow3 = 184,
    #[strum(ascii_case_insensitive)]
    DarkSeaGreen2 = 157,
    #[strum(ascii_case_insensitive)]
    LightCyan3 = 152,
    #[strum(ascii_case_insensitive)]
    LightSkyBlue1 = 153,
    #[strum(ascii_case_insensitive)]
    GreenYellow = 154,
    #[strum(ascii_case_insensitive)]
    DarkOliveGreen2 = 155,
    #[strum(ascii_case_insensitive)]
    DarkSeaGreen1 = 193,
    #[strum(ascii_case_insensitive)]
    PaleTurquoise1 = 159,
    #[strum(ascii_case_insensitive)]
    DeepPink3 = 162,
    #[strum(ascii_case_insensitive)]
    Magenta2 = 200,
    #[strum(ascii_case_insensitive)]
    HotPink2 = 169,
    #[strum(ascii_case_insensitive)]
    Orchid = 170,
    #[strum(ascii_case_insensitive)]
    MediumOrchid1 = 207,
    #[strum(ascii_case_insensitive)]
    Orange3 = 172,
    #[strum(ascii_case_insensitive)]
    LightPink3 = 174,
    #[strum(ascii_case_insensitive)]
    Pink3 = 175,
    #[strum(ascii_case_insensitive)]
    Plum3 = 176,
    #[strum(ascii_case_insensitive)]
    Violet = 177,
    #[strum(ascii_case_insensitive)]
    LightGoldenrod3 = 179,
    #[strum(ascii_case_insensitive)]
    Tan = 180,
    #[strum(ascii_case_insensitive)]
    MistyRose3 = 181,
    #[strum(ascii_case_insensitive)]
    Thistle3 = 182,
    #[strum(ascii_case_insensitive)]
    Plum2 = 183,
    #[strum(ascii_case_insensitive)]
    Khaki3 = 185,
    #[strum(ascii_case_insensitive)]
    LightGoldenrod2 = 222,
    #[strum(ascii_case_insensitive)]
    LightYellow3 = 187,
    #[strum(ascii_case_insensitive)]
    Grey84 = 188,
    #[strum(ascii_case_insensitive)]
    LightSteelBlue1 = 189,
    #[strum(ascii_case_insensitive)]
    Yellow2 = 190,
    #[strum(ascii_case_insensitive)]
    DarkOliveGreen1 = 192,
    #[strum(ascii_case_insensitive)]
    Honeydew2 = 194,
    #[strum(ascii_case_insensitive)]
    LightCyan1 = 195,
    #[strum(ascii_case_insensitive)]
    Red1 = 196,
    #[strum(ascii_case_insensitive)]
    DeepPink2 = 197,
    #[strum(ascii_case_insensitive)]
    DeepPink1 = 199,
    #[strum(ascii_case_insensitive)]
    Magenta1 = 201,
    #[strum(ascii_case_insensitive)]
    OrangeRed1 = 202,
    #[strum(ascii_case_insensitive)]
    IndianRed1 = 204,
    #[strum(ascii_case_insensitive)]
    HotPink = 206,
    #[strum(ascii_case_insensitive)]
    DarkOrange = 208,
    #[strum(ascii_case_insensitive)]
    Salmon1 = 209,
    #[strum(ascii_case_insensitive)]
    LightCoral = 210,
    #[strum(ascii_case_insensitive)]
    PaleVioletRed1 = 211,
    #[strum(ascii_case_insensitive)]
    Orchid2 = 212,
    #[strum(ascii_case_insensitive)]
    Orchid1 = 213,
    #[strum(ascii_case_insensitive)]
    Orange1 = 214,
    #[strum(ascii_case_insensitive)]
    SandyBrown = 215,
    #[strum(ascii_case_insensitive)]
    LightSalmon1 = 216,
    #[strum(ascii_case_insensitive)]
    LightPink1 = 217,
    #[strum(ascii_case_insensitive)]
    Pink1 = 218,
    #[strum(ascii_case_insensitive)]
    Plum1 = 219,
    #[strum(ascii_case_insensitive)]
    Gold1 = 220,
    #[strum(ascii_case_insensitive)]
    NavajoWhite1 = 223,
    #[strum(ascii_case_insensitive)]
    MistyRose1 = 224,
    #[strum(ascii_case_insensitive)]
    Thistle1 = 225,
    #[strum(ascii_case_insensitive)]
    Yellow1 = 226,
    #[strum(ascii_case_insensitive)]
    LightGoldenrod1 = 227,
    #[strum(ascii_case_insensitive)]
    Khaki1 = 228,
    #[strum(ascii_case_insensitive)]
    Wheat1 = 229,
    #[strum(ascii_case_insensitive)]
    Cornsilk1 = 230,
    #[strum(ascii_case_insensitive)]
    Grey100 = 231,
    #[strum(ascii_case_insensitive)]
    Grey3 = 232,
    #[strum(ascii_case_insensitive)]
    Grey7 = 233,
    #[strum(ascii_case_insensitive)]
    Grey11 = 234,
    #[strum(ascii_case_insensitive)]
    Grey15 = 235,
    #[strum(ascii_case_insensitive)]
    Grey19 = 236,
    #[strum(ascii_case_insensitive)]
    Grey23 = 237,
    #[strum(ascii_case_insensitive)]
    Grey27 = 238,
    #[strum(ascii_case_insensitive)]
    Grey30 = 239,
    #[strum(ascii_case_insensitive)]
    Grey35 = 240,
    #[strum(ascii_case_insensitive)]
    Grey39 = 241,
    #[strum(ascii_case_insensitive)]
    Grey42 = 242,
    #[strum(ascii_case_insensitive)]
    Grey46 = 243,
    #[strum(ascii_case_insensitive)]
    Grey50 = 244,
    #[strum(ascii_case_insensitive)]
    Grey54 = 245,
    #[strum(ascii_case_insensitive)]
    Gray58 = 246,
    #[strum(ascii_case_insensitive)]
    Gray62 = 247,
    #[strum(ascii_case_insensitive)]
    Gray66 = 248,
    #[strum(ascii_case_insensitive)]
    Gray70 = 249,
    #[strum(ascii_case_insensitive)]
    Gray74 = 250,
    #[strum(ascii_case_insensitive)]
    Gray78 = 251,
    #[strum(ascii_case_insensitive)]
    Gray82 = 252,
    #[strum(ascii_case_insensitive)]
    Gray85 = 253,
    #[strum(ascii_case_insensitive)]
    Gray89 = 254,
    #[strum(ascii_case_insensitive)]
    Gray93 = 255,
}
pub fn find_color(color: &str) -> i32 {
    let color_variant = Colors::from_str(color.to_case(Case::Pascal).as_str()).unwrap();
    color_variant as i32
}


src/main.rs
use richterm::print;
use richterm::text;
use richterm::track;
use richterm::progress;
use richterm::spinner;

use std::{thread, time};

fn main() {
    let vec = vec![text(
        "richterm",
        "fg:dark_blue bg:indian_red1 eff:b,i,u,blink,s",
    )];
    print(vec);

    let arr = [text(" supports ", "")];
    print(arr);

    print([
        text(" many ", "bg:sea_green2 fg:black eff:i"),
        text(" features ", ""),
        text(" out of the box", "bg:yellow fg:black"),
        text("\n", ""),
    ]);

    print([
        text("Even", ""),
        text(" some ", "eff:b"),
        text(
            "emojis like :+1: :apple: :bar_chart: :airplane_departure: :baguette_bread: :minibus:",
            "",
        ),
        text("\n", ""),
    ]);

    for _item in track(5, "Downloading ...") {
        let ten_millis = time::Duration::from_millis(5);
        thread::sleep(ten_millis);
    }
    
    let mut tasks = progress();
    let task1 = tasks.add_task("Descr1", 50);
    let task2 = tasks.add_task("Descr2", 50); 
    let task3 = tasks.add_task("Descr3", 50);
    
    while !tasks.finished{
        tasks.update(task1.clone(), 10.0);
        tasks.update(task2.clone(), 5.0);
        tasks.update(task3.clone(), 6.0);
    } 

    // let somejob = spinner("moon", "doing job1");
    // somejob.start();
    // // somejob.log([
    // //     text("oh oh", "")
    // //     ]);
    // let ten_millis = time::Duration::from_millis(5000);
    // thread::sleep(ten_millis);
    // somejob.stop();
}
